odl/solvers/smooth/newton.py

```diff
@@ -0,0 +1,380 @@
+
+    if not callable(line_search):
+        line_search = ConstantLineSearch(line_search)
+
+    if cg_iter is None:
+        # Motivated by that if it is Ax = b, x and b in Rn, it takes at most n
+        # iterations to solve with cg
+        cg_iter = grad.domain.size
+
+    # TODO: optimize by using lincomb and avoiding to create copies
+    for _ in range(maxiter):
+
+        # Initialize the search direction to 0
+        search_direction = x.space.zero()
+
+        # Compute hessian (as operator) and gradient in the current point
+        hessian = grad.derivative(x)
+        deriv_in_point = grad(x)
+
+        # Solving A*x = b for x, in this case f''(x)*p = -f'(x)
+        # TODO: Let the user provide/choose method for how to solve this?
+        conjugate_gradient(hessian, search_direction,
+                           -deriv_in_point, cg_iter)
+
+        # Computing step length
+        dir_deriv = search_direction.inner(deriv_in_point)
+        if np.abs(dir_deriv) <= tol:
+            return
+
+        step_length = line_search(x, search_direction, dir_deriv)
+
+        # Updating
+        x += step_length * search_direction
+
+        if callback is not None:
+            callback(x)
+
+
+def bfgs_method(f, x, line_search=1.0, maxiter=1000, tol=1e-16, maxcor=None,
+                hessinv_estimate=None, callback=None):
+    """Quasi-Newton BFGS method to minimize a differentiable function.
+
+    Can use either the regular BFGS method, or the limited memory BFGS method.
+
+    Notes
+    -----
+    This is a general and optimized implementation of a quasi-Newton
+    method with BFGS update for solving a general unconstrained
+    optimization problem
+
+        :math:`\min f(x)`
+
+    for a differentiable function
+    :math:`f: \mathcal{X}\\to \mathbb{R}` on a Hilbert space
+    :math:`\mathcal{X}`. It does so by finding a zero of the gradient
+
+        :math:`\\nabla f: \mathcal{X} \\to \mathcal{X}`.
+
+    The QN method is an approximate Newton method, where the Hessian
+    is approximated and gradually updated in each step. This
+    implementation uses the rank-one BFGS update schema where the
+    inverse of the Hessian is recalculated in each iteration.
+
+    The algorithm is described in [GNS2009]_, Section 12.3 and in the
+    `BFGS Wikipedia article
+    <https://en.wikipedia.org/wiki/Broyden%E2%80%93Fletcher%E2%80%93\
+Goldfarb%E2%80%93Shanno_algorithm>`_
+
+    Parameters
+    ----------
+    f : `Functional`
+        Functional with ``f.gradient``
+    x : ``f.domain`` element
+        Starting point of the iteration
+    line_search : float or `LineSearch`, optional
+        Strategy to choose the step length. If a float is given, uses it as a
+        fixed step length.
+    maxiter : int, optional
+        Maximum number of iterations.
+        ``tol``.
+    tol : float, optional
+        Tolerance that should be used for terminating the iteration.
+    maxcor : int, optional
```

Comment:
Name makes no sense until one reads this doc. Becomes slightly better with an additional "r".
How about using something like "store" in the name to relate it to the storage thing? Like `nstore`, `num_store` or `maxstore`.